### PR TITLE
First Pull request for TCP Connections Analysis in Support Center

### PR DESCRIPTION
**Pull request by puneetg1983**

### CODE REVIEW

**Comment by ShekharGupta1988** [comment]
@puneetg1983 - 
1. I did a merge to master. Looks like after that, there is a branch conflict for this pull request. you would need to resolve that and update the PR
2. This Code Change cannot be deployed until your backend changes are done and deployed, right?

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<availability-graph 
+[(displayGraph)]="displayGraph" >
+</availability-graph>
+
+<h4>TCP Connections Summary</h4>
+<expandable-summary [title]="'TCP Connections'" [summaryViewModel]="connectionsUsageViewModel"></expandable-summary>
+<expandable-summary [title]="'Connections Rejections'" [summaryViewModel]="connnectionsRejectionsViewModel"></expandable-summary>
+<expandable-summary [title]="'Open Socket Handles'" [summaryViewModel]="openSocketCountViewModel"></expandable-summary>
+<br/>
+<feedback-form [source]="'memoryanalysis'"></feedback-form>
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.html]: update the source to be "tcpconnectionanalysis". this source is used for logging purpose

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import { Component, Input, Output, OnInit, EventEmitter } from '@angular/core';
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: Are you using any of these decorators except Component? If not, remove it from there.

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import { ActivatedRoute } from '@angular/router';
+import { IAppAnalysisResponse, IAbnormalTimePeriod, IAnalysisData } from '../../../shared/models/appanalysisresponse';
+import { IDetectorAbnormalTimePeriod, IDetectorResponse } from '../../../shared/models/detectorresponse';
+import { SummaryViewModel, SummaryHealthStatus } from '../../../shared/models/summary-view-model';
+import { PortalActionService, ServerFarmDataService, AvailabilityLoggingService, AuthService, AppAnalysisService } from '../../../shared/services';
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: same for services. I guess lot of these came from a copy command :) . Just remove the ones you are not using

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    templateUrl: 'tcp-connections-analysis.component.html'
+})
+export class TcpConnectionsAnalysisComponent implements OnInit {
+    displayGraph: boolean = true;
+
+    connnectionsRejectionsViewModel: SummaryViewModel;
+    connectionsUsageViewModel: SummaryViewModel;
+    openSocketCountViewModel: SummaryViewModel;
+
+    readonly ConnectionRejections: string = "portexhaustion"
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: missing semicolon ';'

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    readonly TcpConnections: string = "tcpconnectionsusage"
+    readonly OpenSocketCount: string = "tcpopensocketcount"
+
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: remove extra line

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            });
+
+        this.getSummaryViewModel(this.OpenSocketCount, 'TotalOpenSocketCount', false)
+            .subscribe(data => {
+            this.openSocketCountViewModel = data;
+            });
+    }
+
+    getSummaryViewModel(detectorName: string, topLevelSeries: string = '', excludeTopLevelInDetail: boolean = true): Observable<SummaryViewModel> {
+
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: remove extra lines

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            mainGraphDescriptions: 'Connection Rejections is the number of times your application\'s request to open a new connection failed because the machine wide TCP Connection limit was hit',
+            perInstanceGraphTitle: null,
+            perInstanceGraphDescription: null,
+            emptyDataResponse: 'No issues detected with port rejections in this time frame.'
+        },
+        'tcpconnectionsusage': {
+            mainGraphTitle: 'Outbound TCP Connections',
+            mainGraphDescriptions: 'This is the total number of Outbound Connections per Instance',
+            perInstanceGraphTitle: 'TCP Connections',
+            perInstanceGraphDescription: 'The below graph shows connections (both Inbound and Outbound) per instance. Established and TimeWait include both Inbound and Outbound connections'
```
Comment (ShekharGupta1988) [inline app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: did you miss empty Data Response member for tcpconnectionusge?

**Comment by ShekharGupta1988** [comment]
@puneetg1983 - can you put a screenshot for the menu tiles.

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.html (hunk 1 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.html]: done !!!

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 2 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: done

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 3 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: done, yup a result of copy paste :)

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 4 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: done

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 5 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: done

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 6 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: done

File: app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts (hunk 7 above)
Comment (puneetg1983) [reply app/availability/analysis/tcpconnectionsanalysis/tcp-connections-analysis.component.ts]: I was ok with leaving the default message for TCP connections which read something like 'No issues found'

**Comment by puneetg1983** [comment]
Actually this code change need not wait for the deployment at the back-end.  Apart from formatting for the abnormalTimePeriod's message, there isn't any real thing that this pull request depends on from the back-end. I have coded the backend to return a single abnormalTimePeriod and currently the UI is reading the first abnormalTimePeriod anyways like this.

```
return <SummaryViewModel>{
detectorName: detectorName,
health: health,
loading: false,
detectorAbnormalTimePeriod: detectorResponse.abnormalTimePeriods[0],
```

So even if this gets deployed, the only missing this will be the formatted message which is shown above each graph

Here is the screenshot
image

Fixing the conflict, let me know if I missed something else

**Comment by ShekharGupta1988** [comment]
+ @Jen7714 : Hey Jen, can you review the color scheme and title for TCP Connections Tile in home page. See attached screenshot above

**Comment by ShekharGupta1988** [comment]
hi @puneetg1983 - the changes looks good. I am approving the PR. There might be changes needed in the tile color based on Jen's feedback.

**Comment by puneetg1983** [comment]
Done.

**Comment by stevenernst131** [comment]
+ 1 on changing the color of the tile

**Comment by Jen7714** [comment]
Maybe we can change it to one of these: 

#009999
#780780
#39aaf1

**Pull request merged**